Technical changes in logging customization

storojs72 opened the pull request:
These changes are rather technical. 

Main point is soonest customization for logging process (as we discussed by voice - it is critical 1) to preserve format for logging in error cases, even if not all parameters for customization have been already loaded, and 2) to start logging only when format is customized). 

Since format is specified by actual formatter object (`CefFormatter`, `JsonFormatter`, `TextFormatter`), it is necessary to create this object first, and then to perform customization on it. For this reason - `SetFormat(*loggingFormat)` function (that previously returns nothing) is changed onto function `logging.CreateFormatter(*loggingFormat)` that returns actual formatter object with empty fields. Then, as long as parameters become ready (basically hooks and service name - which is actually known but applying it requires formatter object whatever) - those fields are initialised. 

Mentioned "approach" is applied for each service instead of previous customization via builder

ilammy commented on cmd/acra-connector/acra-connector.go:
```diff
@@ -255,7 +253,11 @@ func main() {
-	logging.Customize().SetFormat(*loggingFormat).SetServiceName(ServiceName).Complete()
+	// Start customizing logs here (directly after command line arguments parsing)
+	formatter := logging.CreateFormatter(*loggingFormat)
+	logging.SetServiceName(formatter, ServiceName)
```
It's weird for **acra-connector** to be the only one without
```go
log.SetOutput(os.Stderr)
```
Is there a reason for that?

storojs72 replied:
no. Just my oversight, thank you

ilammy commented on cmd/acra-translator/acra-translator.go:
```diff
@@ -83,7 +83,10 @@ func main() {
+	// Start customizing logs here (directly after command line arguments parsing)
+	formatter := logging.CreateFormatter(*loggingFormat)
+	logging.SetServiceName(formatter, ServiceName)
+	log.SetOutput(os.Stderr)
```
Similarly, **acra-translator** seems to be the only one without

```go
log.WithField("version", utils.VERSION).Infof("Starting service %v [pid=%v]", ServiceName, os.Getpid())
```

during startup.

storojs72 replied:
thank you

ilammy commented on logging/log_formatters.go:
```diff
@@ -32,55 +32,35 @@ import (
 			QuoteEmptyFields: true,
 		},
-		Hooks: hooks,
+		Hooks: nil,
```
Fields are initialized with `nil` by default so this can be omitted. However, I don't particularly mind explicit initialization either.

Lagovas replied:
omit default values is good to keep fewer lines of code, less time to verify formatting, etc ) but anyway didn't find any suggestions/conventions on [effective go](https://golang.org/doc/effective_go.html), so up to you

ilammy commented on logging/log_formatters.go:
```diff
@@ -126,6 +106,14 @@ type AcraTextFormatter struct {
 
+func (f *AcraTextFormatter) SetServiceName(serviceName string) {
+	// service name is ignored by plaintext formatter, so just do nothing
+}
```
Oh... This is news to me. I thought it is mapped onto a log field or something, but now I've noticed that indeed, it's not really used in any way by the plaintext formatter.

By the way, is there a particular reason for why we have the “service name” thing at all, @Lagovas, @shadinua? Shouldn't the log connector be aware of the log source so that it's not necessary to set it? Or is it only because CEF has a mandatory field for that? Why JSON follows suit then?

Lagovas replied:
it's new for me too) yes, maybe we added it first when needed cef format. but anyway it's good to log source of logs especially when we have several running services to distinguish them

ilammy commented on logging/logging.go:
```diff
@@ -71,6 +69,15 @@ func IsDebugLevel(logger *log.Entry) bool {
+// FormatterWrapper wraps log.Formatter interface and adds functions for customizations.
+// Intention for this interface is to provide ability to customize logging by accustomed:
+// `logging.SetServiceName` / `logging.SetHooks` from main function of Acra services
+type FormatterWrapper interface {
```
I'd call it just `Formatter` since the users don't need to actively *wrap* anything into it. It's just that `logging.Formatter` is a `logrus.Formatter` with a couple of additional methods that we need. Acra formatters already implement this interface.

Nice use of interface for customization, by the way! 👍

ilammy commented on logging/logging.go:
```diff
@@ -84,79 +91,41 @@ func SetLogLevel(level int) {
+	case CefFormatString:
+		formatter = CEFFormatter()
+	default:
+		formatter = TextFormatter()
```
I see that it's the behavior we had before, but is it okay to let the users say `--log_format "cat videos"` and get the default one without any error?

storojs72 replied:
Yes, probably it is better to throw some error like `unexpected logs format has been specified`

Lagovas commented on logging/logging.go:
```diff
@@ -40,9 +38,9 @@ const (
-	LoggingFormatPlaintext = "plaintext"
-	LoggingFormatJSON      = "json"
-	LoggingFormatCEF       = "cef"
+	PlaintextFormatString = "plaintext"
```
why renamed?

storojs72 replied:
Because IDE highlighted those constants for me as they start same as package name

Lagovas commented on logging/logging.go:
```diff
@@ -84,79 +91,41 @@ func SetLogLevel(level int) {
-	return c
+// SetServiceName adds service-name label to log entries
+// (plaintext formatter ignores it)
+func SetServiceName(formatter FormatterWrapper, serviceName string) {
```
why not add method `SetServiceName` to formatter object? Why we use OOP with functions and explicit first argument to object?)

storojs72 replied:
The reason is just stylistic. Previously we used to apply customizations like `log.SetOutput()`, `logging.SetServiceName()`, etc. So I decided to keep this style.

storojs72 replied:
Added https://github.com/cossacklabs/acra/pull/370/commits/4247aed0c2c46a91927634a72af3928a45511183 after @ilammy 's explanation https://github.com/cossacklabs/acra-enterprise/pull/41#discussion_r400994241

storojs72 commented:
- added missing `log.SetOutput` to connector;
- added missing `log.WithField("version", utils.VERSION).Infof("Starting service %v [pid=%v]", ServiceName, os.Getpid())` to translator;
- renamed `FormatterWrapper` -> `Formatter`
